xcc: fix building errors related to logging

dcpleung opened the pull request:
XCC is based on very old GCC (sigh...) that there is no auto type, and _Pragma cannot be used within functions. The latest logging changes related to argument evaluation resulted in build errors regarding these. See individual commits for more details.

dcpleung committed:
logging: v2: workaround for old toolchains when !LOG

For GCC < 4.9.0 and Clang < 3.8.0, auto type is not supported.
But the previous behavior to depend on CONFIG_LOG2_ALWAYS_RUNTIME
to be enabled doesn't work because this kconfig is not available
when CONFIG_LOG is not enabled, as LOG_*() are still being
expanded when CONFIG_LOG=n, resulting in toolchain complaining
about unknown keyword. So when CONFIG_LOG=n and old toolchains,
force it to use runtime packaging to avoid the issue.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
logging: v2: force LOG2_ALWAYS_RUNTIME if XCC

XCC is based on GCC 4.2.0 which doesn't support auto type.
So force CONFIG_LOG2_ALWAYS_RUNTIME to be enabled if XCC is
being used.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

cgturner1 commented:
i did test these with XCC for all adsp targets with the issues mentioned in #43087 and now everything builds as expected without the _pragma related errors

andyross commented on include/logging/log_msg2.h:
```diff
@@ -34,6 +35,12 @@ typedef uint64_t log_timestamp_t;
 
+#if !defined(CONFIG_LOG) && \
+	!GCC_VERSION_AT_LEAST(4, 9, 0) && \
+	!CLANG_VERSION_AT_LEAST(3, 8, 0)
```
Nitpick: abstraction-wise, might be better to have toolchain.h define "TOOLCHAIN_HAS_AUTO_TYPE" or somesuch instead of putting the logic here in logging.

dcpleung replied:
It also guards against using `_Pragma()` within functions, so not exactly about auto type only.

nordic-krch replied:
i'm also for moving flags to toolchain.h.
`TOOLCHAIN_HAS_AUTO_TYPE`, `TOOLCHAIN_HAS_C11_GENERIC`, maybe `TOOLCHAIN_HAS_DIAGNOSTIC_PRAGMAS` (seems that this was introduced in 4.6).

It's already third place i'm aware of where we do that (one in PR https://github.com/zephyrproject-rtos/zephyr/blob/9705d17e11cd119269d3e7ddba505af8dac0d56b/include/event_manager/event_manager_priv.h#L33).

andyross commented on include/logging/log_msg2.h:
```diff
@@ -388,8 +395,11 @@ do { \
 
+#ifdef LOG_DISABLED_OLD_TOOLCHAIN_WORKAROUND
+#undef LOG_DISABLED_OLD_TOOLCHAIN_WORKAROUND
+#endif
```
Why undef it?  Not understanding this part.  This header has an include guard anyway, so it's not like you're going to have variant/duplicate definitions of this floating around a compilation unit.

dcpleung replied:
In general, just like to clean up things that should not be used after this.

andyross reviewed (dismissed):
Some nitpicks but nothing worth blocking merge over

dcpleung committed:
toolchain: introduce GCC_VERSION macro

Adds a GCC_VERSION macro to make it easier to write
preprocessor conditionals.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
toolchain: introduce CLANG_VERSION macro

Adds a CLANG_VERSION macro to make it easier to write
preprocessor conditionals.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
toolchain: introduce macro TOOLCHAIN_HAS_PRAGMA_DIAG

This introduces the macro TOOLCHAIN_HAS_PRAGMA_DIAG to indicate
that the toolchain supports #pragma diagnostics. This is
supported since GCC 4.6.0 and, AFAIK, all Clang.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
toolchain: introduce macro TOOLCHAIN_HAS_C_GENERIC

This introduces the macro TOOLCHAIN_HAS_C_GENERIC to indicate
that the toolchain supports C Generic (i.e. _Generic keyword).
This is introduced in C11, and is also supported since GCC 4.9.0
and Clang 3.8.0.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
toolchain: introduce macro TOOLCHAIN_HAS_C_AUTO_TYPE

This introduces the macro TOOLCHAIN_HAS_C_AUTO_TYPE to
indicate that the toolchain supports __auto_type. This is
supported since GCC 4.9.0 and Clang 3.8.0.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>
